internal/r8/deobfuscator.go

```diff
@@ -0,0 +1,205 @@
+	var res = make(map[string]string)
+	scanner := bufio.NewScanner(mapReader)
+	var currentType *MappedType
+	var currentMappedMethodCall *MappedMethodCall
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		typeMatch := typePattern.FindStringSubmatch(line)
+		if typeMatch != nil {
+			currentMappedMethodCall = nil
+			if currentType != nil {
+				mapLeftoverUnmappedMethods(res, currentType)
+			}
+			obfuscatedName := typeMatch[2]
+			stacktraceType, ok := symbols[obfuscatedName]
+			if ok {
+				currentType = &MappedType{stacktraceType, typeMatch[1]}
+			} else {
+				currentType = nil
+			}
+		} else if currentType != nil {
+			methodMatch := methodPattern.FindStringSubmatch(line)
+			if methodMatch != nil {
+				currentMappedMethodCall = upsertMappedMethodCall(res, MethodMatch{
+					methodMatch[1],
+					methodMatch[2],
+					methodMatch[3],
+					methodMatch[4],
+				}, currentType, currentMappedMethodCall)
+			}
+		}
+	}
+	if currentType != nil {
+		mapLeftoverUnmappedMethods(res, currentType)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// Checks if the found methodMatch from the map file is part of the methods previously parsed for the currentType within the
+// stacktrace.
+// If the method is found, then it's added to the mapping in order to be used for replacing the obfuscated name later, it also returns it as the current MappedMethodCall.
+// If the method is NOT found, but it turns out to be a continuation of a previously found method (currentMappedMethodCall), then it appends it to the existing method replacement in the mapping.
+// If the method is NOT found and is also NOT a continuation for the currentMappedMethodCall, then the methodMatch is ignored.
+func upsertMappedMethodCall(mapping map[string]string, methodMatch MethodMatch, currentType *MappedType, currentMappedMethodCall *MappedMethodCall) *MappedMethodCall {
+	methodNameReference := methodMatch.methodObfuscatedName
+	if methodMatch.sourceFileStart != "" {
+		if methodMatch.sourceFileStart != methodMatch.sourceFileEnd {
+			// This is probably due an edge-case where the mapping line starts with different numbers (e.g 1:2). We don't
+			// have that case in our tests, therefore we are ignoring it.
+			return currentMappedMethodCall
+		}
+		methodNameReference = fmt.Sprintf("%s:%s", methodMatch.methodObfuscatedName, methodMatch.sourceFileStart)
+	}
+	mapReference := currentType.obfuscated.name + ":" + methodNameReference
+	methodCallSite, ok := currentType.obfuscated.methods[methodNameReference]
+	if ok {
+		// Found this method in the list of methods parsed from the stacktrace for the currentType.
+		delete(currentType.obfuscated.methods, methodNameReference)
+		key := getKey(currentType.obfuscated.name, methodMatch.methodObfuscatedName, methodCallSite)
+		mapping[key] = getKey(currentType.realName, methodMatch.methodRealName, methodCallSite)
+		return &MappedMethodCall{mapReference, key}
```

Comment:
nit: minor readability improvement
```suggestion
		return &MappedMethodCall{reference: mapReference, key: key}
```